src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy

```diff
@@ -0,0 +1,261 @@
+import javax.ws.rs.core.Response
+import javax.ws.rs.core.UriInfo
+import java.util.regex.Pattern
+
+@Path("/locations")
+@Produces(MediaType.APPLICATION_JSON)
+class LocationResource {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LocationResource.class)
+    public static final ArrayList<String> ALLOWED_CAMPUSES = ["corvallis"]
+    public static final ArrayList<String> ALLOWED_TYPES = ["building", "dining"]
+
+    /**
+     * Default page number used in pagination
+     * @todo: should this be coming from skeleton or configuration file?
+     */
+    public static final Integer DEFAULT_PAGE_NUMBER = 1
+
+    /**
+     * Default page size used in pagination
+     * @todo: should this be coming from skeleton or configuration file?
+     */
+    public static final Integer DEFAULT_PAGE_SIZE = 10
+
+    @Context
+    UriInfo uriInfo
+
+    private final Map<String, String> locationConfiguration
+
+    private static final Pattern illegalCharacterPattern = Pattern.compile(
+            '''(?x)       # this extended regex defines
+               (?!        # any character that is not
+                  [
+                   a-zA-Z # a letter,
+                   0-9    # a number,
+                   -      # a hyphen,
+                   _      # an underscore,
+                   \\.    # a period, or
+                   @      # an at sign
+                  ])
+               .          # to be an illegal character.
+            ''')
+
+    LocationResource(Map<String, String> locationConfiguration) {
+        this.locationConfiguration = locationConfiguration
+    }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Timed
+    Response list(@QueryParam('q') String q, @QueryParam('campus') String campus, @QueryParam('type') String type,
+                  @Auth AuthenticatedUser authenticatedUser) {
+
+        def pageNumber = getPageNumber()
+        def pageSize = getPageSize()
+        def trimmedQ = sanitize(q?.trim())
+        def trimmedCampus = sanitize(campus?.trim())
+        def trimmedType = sanitize(type?.trim())
+
+        if (trimmedCampus && !ALLOWED_CAMPUSES.contains(trimmedCampus)) {
+            throw new WebApplicationException(Response.Status.NOT_FOUND)
+        }
+
+        if (trimmedType && !ALLOWED_TYPES.contains(trimmedType)) {
+            throw new WebApplicationException(Response.Status.NOT_FOUND)
+        }
+
+        def query = [ "match_all": [:] ]
+        def esFullUrl = getESFullUrl()
+
+        def esQuery = getESSearchQuery(trimmedCampus, trimmedType, trimmedQ, pageNumber, pageSize, query)
+        ObjectMapper mapper = new ObjectMapper(); // can reuse, share globally
+        String esQueryJson = mapper.writeValueAsString(esQuery)
+
+        def url = new URL(esFullUrl + "/_search")
+        URLConnection connection = postRequest(url, esQueryJson)
+
+        def esReponse = connection.content.text
+
+        Response.ok(esReponse).build()
+    }
```

Comment:
You are right. This class is doing too much. I wanted to see what the final result was like before refactoring things out.